TrySingle

jimmcslim opened the pull request:
Implementation of `TrySingle` that I proposed some time ago in #540.

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    partial class MoreEnumerable
```
Add line break before for visual separation from imports, as in:

```suggestion

    partial class MoreEnumerable
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+    partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Similar to Single() or SingleOrDefault() but allows the caller to determine
```
Summaries should not refer to other methods. If they do, use `<see cref="..." />` to call them out.

jimmcslim replied:
Yeah, I normally do that but was getting a compiler error. I'm sure the imports were valid as well...

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <typeparam name="TCardinality">The type that expresses cardinality.</typeparam>
+        /// <typeparam name="TResult">The result type of the resultSelector.</typeparam>
+        /// <returns>The value provided by the resultSelector.</returns>
+        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> values,
```
Rename `values` to `source` (and fix-up doc summary accordingly):

```suggestion
        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// Similar to Single() or SingleOrDefault() but allows the caller to determine
+        /// whether there were zero or many elements in the sequence in the event that there isn't just one.
+        /// </summary>
+        /// <param name="values">The source sequence that will be tested for its cardinality.</param>
```
Rename to `source`:

```suggestion
        /// <param name="sources">The source sequence that will be tested for its cardinality.</param>
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// whether there were zero or many elements in the sequence in the event that there isn't just one.
+        /// </summary>
+        /// <param name="values">The source sequence that will be tested for its cardinality.</param>
+        /// <param name="zero">The value that should be provided to resultSelector if the sequence has zero elements.</param>
```
Use `paramref` to refer to parameter, as in:

```suggestion
        /// <param name="zero">The value that should be provided to <paramref name="resultSelector" /> if the sequence has zero elements.</param>
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// </summary>
+        /// <param name="values">The source sequence that will be tested for its cardinality.</param>
+        /// <param name="zero">The value that should be provided to resultSelector if the sequence has zero elements.</param>
+        /// <param name="one">The value that should be provided to resultSelector if the sequence has one element.</param>
```
Use `paramref` to refer to parameter, as in:

```suggestion
        /// <param name="one">The value that should be provided to <paramref name="resultSelector" /> if the sequence has one element.</param>
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <param name="values">The source sequence that will be tested for its cardinality.</param>
+        /// <param name="zero">The value that should be provided to resultSelector if the sequence has zero elements.</param>
+        /// <param name="one">The value that should be provided to resultSelector if the sequence has one element.</param>
+        /// <param name="many">The value that should be provided to resultSelector if the sequence has two or more elements.</param>
```
Use `paramref` to refer to parameter, as in:

```suggestion
        /// <param name="many">The value that should be provided to <paramref name="resultSelector" /> if the sequence has two or more elements.</param>
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <param name="many">The value that should be provided to resultSelector if the sequence has two or more elements.</param>
+        /// <param name="resultSelector">A function that is provided with the cardinality, and if the sequence has just
+        /// one element, the value of that element. Then transforms the result to an instance of TResult.</param>
+        /// <typeparam name="T">The type of the elements of the sequence</typeparam>
```
End in a period/full-stop:

```suggestion
        /// <typeparam name="T">The type of the elements of the sequence.</typeparam>
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// one element, the value of that element. Then transforms the result to an instance of TResult.</param>
+        /// <typeparam name="T">The type of the elements of the sequence</typeparam>
+        /// <typeparam name="TCardinality">The type that expresses cardinality.</typeparam>
+        /// <typeparam name="TResult">The result type of the resultSelector.</typeparam>
```
```suggestion
        /// <typeparam name="TResult">The result type returned by <paramref name="resultSelector" />.</typeparam>
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+        /// <typeparam name="T">The type of the elements of the sequence</typeparam>
+        /// <typeparam name="TCardinality">The type that expresses cardinality.</typeparam>
+        /// <typeparam name="TResult">The result type of the resultSelector.</typeparam>
+        /// <returns>The value provided by the resultSelector.</returns>
```
```suggestion
        /// <returns>The value returned by the <paramref name="resultSelector" />.</returns>
```

atifaziz replied:
?

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,58 @@
+            if (values == null) throw new ArgumentNullException(nameof(values));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            var result = values.Take(2).ToArray();
```
We can optimise after but it would be nice to just do it now by avoiding conversion to array to just have convenient access to the array's length to determine the result.

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            var arrayWithNone = new int?[0];
+
+            var result = arrayWithNone.TrySingle("zero", "one", "many", ((cardinality, value) => (cardinality, value)));
```
Consider using deconstruction here instead. I'm suggesting:

```c#
var (cardinality, value) = arrayWithNone.TrySingle("zero", "one", "many", ValueTuple.Create);
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,56 @@
+    public class TrySingleTest
+    {
+        [Test]
+        public void TrySingleWithNone()
```
Rename to:

```suggestion
        public void TrySingleWithEmptySource()
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        [Test]
+        public void TrySingleWithOne()
```
Rename to:

```suggestion
        public void TrySingleWithSingleton()
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,56 @@
+        [Test]
+        public void TrySingleWithOne()
+        {
+            var arrayWithOne = new int?[] {10};
```
```suggestion
            var arrayWithOne = new int?[] { 10 };
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            var arrayWithOne = new int?[] {10};
+
+            var result = arrayWithOne.TrySingle("zero", "one", "many", (cardinality, value) => (cardinality, value));
```
Ditto as earlier comment about deconstruction.

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            var arrayWithMultiple = new int?[] {10, 20};
+
+            var result = arrayWithMultiple.TrySingle("zero", "one", "many", (cardinality, value) => (cardinality, value));
```
Ditto as earlier comment about deconstruction.

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,56 @@
+        [Test]
+        public void TrySingleWithMoreThanOne()
+        {
+            var arrayWithMultiple = new int?[] {10, 20};
```
```suggestion
            var arrayWithMultiple = new int?[] { 10, 20 };
```

jimmcslim commented:
> Apart from other review comment, you need to run `build.cmd` locally before committing to fix the [issue of synchronizing the generated wrappers](https://ci.appveyor.com/project/raboof/morelinq/builds/27958526#L494).
> 
> Thanks!

I'll address the rest of these this evening.

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,61 @@
+        /// <typeparam name="TCardinality">The type that expresses cardinality.</typeparam>
+        /// <typeparam name="TResult">The result type of the resultSelector.</typeparam>
+        /// <returns>The value provided by the resultSelector.</returns>
+        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```
Add a blank line between  doc summary and method declaration for clarity:
```suggestion

                public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```

jimmcslim replied:
Whoops, just overlooked. Committed.

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,62 @@
+        /// <typeparam name="TResult">The result type of the resultSelector.</typeparam>
+        /// <returns>The value provided by the resultSelector.</returns>
+
+                public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```
Odd indentation level. Suggested fix:

```suggestion
        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,62 @@
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            var result = source.Take(2).ToArray();
```
How about simplifying the implementation to the following?

```c#
            using (var e = source.GetEnumerator())
            {
                return !e.MoveNext() ? resultSelector(zero, default)
                     : !e.MoveNext() ? resultSelector(one, e.Current)
                     : resultSelector(many, default);
            }
```

This avoids two extra allocations with the call to `ToArray()`.

jimmcslim replied:
I'll take a look this evening.

jimmcslim replied:
I've updated it to something similar; I found on Mono that `e.Current` can't be called after the second call to `e.MoveNext()`, so I pull `e.Current` into a variable, before checking if there are more elements in the sequence.

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,79 @@
+    partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Determines the cardinality of the source sequence, provides a
```
This summary needs re-phrasing to closely align with `TrySingle` as the name of the method. It should not start by saying that it determines the cardinality of the sequence but rather that it _attempts_ to retrieve and project a result from a sequence of one element; and failing that, it provides the cardinality of the sequence for the zero or many cases.

jimmcslim replied:
I've updated, but had to put a fairly ugly and long `<see cref=""/>` to `Enumerable.Single()` to explain the method.

jimmcslim commented:
I've added optimisation for `ICollection<T>` and `IList<T>`.

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,132 @@
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using MoreLinq.Extensions;
```
This import seems unused.

```suggestion
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        [Test]
+        public void TrySingleOptimizesForIList()
```
Let's refer to a list as simply a _list_ as opposed to spelling out the interface:

```suggestion
        public void TrySingleOptimizesForList()
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,111 @@
+        {
+            var arrayWithMultiple = new int?[] { 10, 20 };
+
+            var (cardinality, value) = arrayWithMultiple.TrySingle("zero", "one", "many", ValueTuple.Create);
```
Let's use the simpler overload:

```suggestion
            var (cardinality, value) = arrayWithMultiple.TrySingle("zero", "one", "many");
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,111 @@
+        {
+            var arrayWithNone = new int?[0];
+
+            var (cardinality, value) = arrayWithNone.TrySingle("zero", "one", "many", ValueTuple.Create);
```
Let's use the simpler overload:

```suggestion
            var (cardinality, value) = arrayWithNone.TrySingle("zero", "one", "many");
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,111 @@
+            {
+                yield return 1;
+                yield return 2;
+                throw new Exception("TrySingle should not have attempted to consume a third element.");
```
Might as well use `nameof` since we can:

```suggestion
                throw new Exception(nameof(MoreEnumerable.TrySingle) + " should not have attempted to consume a third element.");
```

Or even do without mentioning the method under test:

```suggestion
                Assert.Fail("An unnecessary attempt was made to extract the third element.");
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,111 @@
+            Assert.AreEqual("zero", cardinality);
+            Assert.AreEqual(default(int), value);
+
+            list = new BreakingList<int>(new List<int> {1});
```
```suggestion
            list = new BreakingList<int>(new List<int> { 1 });
```

jimmcslim commented:
Ok, have responded to the feedback, implemented the additional tests required, and refactored. Couldn't use `BreakingCollection`  and `BreakingReadOnlyCollection` in the single element case, so have implemented my own `BreakingSingleElementCollection` and `BreakingSingleElementReadOnlyCollection` for my testing requirements.

jimmcslim commented:
I've also added an overload in branch `overload_proposal` of my fork, that is similar to my original proposal `TCardinality TrySingle<T, TCardinality>(this IEnumerable<T> source, TCardinality zero, TCardinality one, TCardinality many, out T result)` - do you think this is appropriate to include?

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,171 @@
+        /// <see cref="IList{T}"/> or <see cref="ICollection{T}"/> then the
+        /// implementation optimizes by checking the number of elements in
+        /// the underlying sequence.</remarks>
+        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```
Add a blank line between doc summary and method declaration for clarity:

```suggestion

        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,171 @@
+        /// <see cref="IList{T}"/> or <see cref="ICollection{T}"/> then the
+        /// implementation optimizes by checking the number of elements in
+        /// the underlying sequence.</remarks>
+        public static (TCardinality Cardinality, T Value)
```
Add a blank line between doc summary and method declaration for clarity:

```suggestion

        public static (TCardinality Cardinality, T Value)
```

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,171 @@
+                        return resultSelector(zero, default);
+                    case 1:
+                        return resultSelector(one,
+                            theCollection is IReadOnlyList<T> theList ? theList[0]
```
I know it's odd to say this, but just because a type implements `IReadOnlyCollection<>` doesn't mean it will implement `IReadOnlyList<>` and not `IList<>`. One should cater for both.

atifaziz commented on MoreLinq/TrySingle.cs:
```diff
@@ -0,0 +1,171 @@
+                    return TrySingleForEnumerable(source);
+            }
+
+            TResult TrySingleForReadOnlyCollection(IReadOnlyCollection<T> theCollection)
```
This is largely redundant with `TrySingleForCollection`. I think you could just fold all the local functions into a single 3-level switch that it really is. Like this?

```c#
            switch (source.TryGetCollectionCount())
            {
                case int count:
                {
                    switch (count)
                    {
                        case 0:
                            return resultSelector(zero, default);
                        case 1:
                            T item;
                            switch (source)
                            {
                                case IReadOnlyList<T> list: item = list[0]; break;
                                case IList<T> list: item = list[0]; break;
                                default: item = source.First(); break;
                            }
                            return resultSelector(one, item);
                        default:
                            return resultSelector(many, default);
                    }
                }
                default:
                {
                    using (var e = source.GetEnumerator())
                    {
                        if (!e.MoveNext())
                            return resultSelector(zero, default);
                        var current = e.Current;
                        return !e.MoveNext() ? resultSelector(one, current)
                                             : resultSelector(many, default);
                    }
                }
            }
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+        }
+
+        [Test]
+        public void TrySingleOptimizesForCollection()
```
Isn't this redundant/obsolete now?

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+        }
+
+        [Test]
+        public void TrySingleOptimizesForList()
```
Isn't this redundant/obsolete now?

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+            Assert.AreEqual(value, result);
+        }
+
+        private static object[] _singletonCollectionTestCases =
```
- Remove `private` that is redundant in C#
- Add `readonly` modifier
- Rename to more conventional `SingletonCollectionTestCases` (Pascal casing is used for static read-only fields even if private)
- Change field type to `ITestCaseData[]`, which is strong than `object[]`

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+
+        private static object[] _singletonCollectionTestCases =
+        {
+            new object[] { new BreakingSingleElementCollection<int>(10), 10 },
```
Given above, this together with the next should instead use instantiations of `TestCaseData`:

```c#
            new TestCaseData(new BreakingSingleElementCollection<int>(10), 10),
            new TestCaseData(new BreakingSingleElementReadOnlyCollection<int>(20), 20),
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+
+        private class BreakingSingleElementReadOnlyCollection<T> : IReadOnlyCollection<T>
+        {
+            private readonly T _element;
```
```suggestion
            readonly T _element;
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+            public bool IsReadOnly => true;
+        }
+
+        private class BreakingSingleElementReadOnlyCollection<T> : IReadOnlyCollection<T>
```
```suggestion
        sealed class BreakingSingleElementReadOnlyCollection<T> : IReadOnlyCollection<T>
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+            }
+        }
+
+        private class BreakingSingleElementCollection<T> : ICollection<T>
```
```suggestion
        sealed class BreakingSingleElementCollection<T> : ICollection<T>
```

atifaziz replied:
Also, this can share its implementation with `BreakingSingleElementReadOnlyCollection` via a base `BreakingSingleElementSequence` to avoid redundancies. Here's what I am thinking:

```c#
class BreakingSingleElementCollectionBase<T> : IEnumerable<T>
{
    readonly T _element;

    protected BreakingSingleElementCollectionBase(T element) => _element = element;

    public int Count => 1;

    public IEnumerator<T> GetEnumerator()
    {
        yield return _element;
        throw new Exception($"{nameof(MoreEnumerable.TrySingle)} should not have attempted to consume a second element.");
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}

sealed class BreakingSingleElementCollection<T> :
    BreakingSingleElementCollectionBase<T>, ICollection<T>
{
    public BreakingSingleElementCollection(T element) : base(element) {}

    public void Add(T item) => throw new NotImplementedException();
    public void Clear() => throw new NotImplementedException();
    public bool Contains(T item) => throw new NotImplementedException();
    public void CopyTo(T[] array, int arrayIndex) => throw new NotImplementedException();
    public bool Remove(T item) => throw new NotImplementedException();
    public bool IsReadOnly => true;
}

sealed class BreakingSingleElementReadOnlyCollection<T> :
    BreakingSingleElementCollectionBase<T>, IReadOnlyCollection<T>
{
    public BreakingSingleElementReadOnlyCollection(T element) : base(element) {}
}
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+
+        private class BreakingSingleElementCollection<T> : ICollection<T>
+        {
+            private readonly T _element;
```
```suggestion
            readonly T _element;
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            private readonly T _element;
+
+            public BreakingSingleElementCollection(T element)
```
Rewrite more simply as expression-bodied member, as in:

```c#
public BreakingSingleElementCollection(T element) => _element = element;
```

atifaziz commented on MoreLinq.Test/TrySingleTest.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            private readonly T _element;
+
+            public BreakingSingleElementReadOnlyCollection(T element)
```
Rewrite more simply as expression-bodied member.

jimmcslim commented:
No worries, thanks for your reviews. I appreciate your eye toward high-quality in this library and it has been great to learn some more about the patterns that it uses.
